### PR TITLE
feat(auth): Add access control checks to all mutations [BACK-1311]

### PR DESCRIPTION
**Pull request by nina-py**
## Goal

Add access checks so that only users in the `mozilliansorg_pocket_collection_curator_full` can execute mutations on collections & collections-related entities. Since all mutation resolvers in this repository use a single helper function `executeMutation`, it was very easy to add access checks for all mutations at once in one place.

Note that this repository has full test coverage at the database resolver level but only partial coverage for Apollo integration tests. The Apollo integration tests, located at `src/test/admin-server/mutations`, only cover mutations for `CollectionAuthor` and `CollectionPartner`. I have updated these tests and added a new test for each mutation to check the access denied error. 

Given that all mutations run the same access checking code, it is safe to presume other mutations will also not be executed if the user doesn't have permission to do so. However, we should at least add writing the outstandng Apollo integration tests to the backlog somewhere (@jpetto - what do you think?).

## Reference

Tickets:

- https://getpocket.atlassian.net/browse/BACK-1311

### CODE REVIEW

File: src/test/admin-server/index.ts [hunk 1]
```diff
@@ -1,25 +1,51 @@
...
+          },
+          db: client(),
+          s3,
+        })
+      );
+    },
+    // Note the absence of the Sentry plugin - it emits
+    // "Cannot read property 'headers' of undefined" errors in tests.
+    // We get console.log statements that resolvers emit instead
+    // but the tests pass.
```
Comment (Herraj) [inline src/test/admin-server/index.ts:44]: Could this be fixed by passing in the `req` object to our context function?
i.e:
```
 context: ({req}) => {
      // If context has been provided, use that instead.
      return (
        context ??
        new ContextManager({
          request: {
            ...req,
            headers: {},
          },
          ...
```

File: src/admin/resolvers/mutations.ts [hunk 2]
```diff
@@ -50,36 +51,43 @@ import {
...
+    const { db, authenticatedUser } = context;
 
-    // Associate the image with the entity if the image entity type is provided
-    // and a record for the image exists
-    if (imageEntityType) {
-      await associateImageWithEntity(
-        db,
-        entity as U & { id: number; imageUrl: string },
-        imageEntityType
+    if (authenticatedUser.hasFullAccess) {
```
Comment (Herraj) [inline src/admin/resolvers/mutations.ts:73]: We could also flip this and check e.g `! authenticatedUser.hasFullAccess` and throw the error and do an early exit but this is fine too

**Review by Herraj — COMMENTED** [review]
Good work! 

I'd just like one addition which is we add a test or two for when the `headers` is `undefined`. I didn't have it initially in my prospect api auth work but I added it later on which could've helped me catch a bug earlier.

Also in hindsight, what do you think about creating a separate `describe( ...)` for tests that test the _unauthorized_ requests? That way we can set up unauthorized server headers once instead of copying new ones in each test. Just a suggestion, I'm not sure how much refactoring you'll have to do. But I guess you could just get by doing this in each test:

`const server = getServerWithMockedHeaders({...headers, groups:'no-access-for-uo'}); `

File: src/main.ts [hunk 3]
```diff
@@ -5,8 +5,9 @@ import xrayExpress from 'aws-xray-sdk-express';
 import express from 'express';
 import https from 'https';
 import { graphqlUploadExpress } from 'graphql-upload';
-import { server as adminServer } from './admin/server';
+import { startAdminServer } from './admin/server';
 import { server as publicServer } from './public/server';
+import { getContext } from './admin/context';
```
Comment (jpetto) [inline src/main.ts:10]: could we be consistent on naming/importing things from `admin`? i think either of these would work:

```
import { startServer as startAdminServer } from './admin/server';
import { getContext as getAdminContext } from './admin/context';
```

```
import { startAdminServer } from './admin/server';
import { getAdminContext } from './admin/context';
```

File: src/admin/server.ts [hunk 4]
```diff
@@ -1,29 +1,52 @@
...
+  (req: Request): ContextManager;
+}
+
+/**
+ * Sets up and configures an ApolloServer for the application.
+ * @param contextFactory function for creating the context with
+ * every request
+ * @returns ApolloServer
+ */
+export function getServer(contextFactory: ContextFactory): ApolloServer {
```
Comment (jpetto) [inline src/admin/server.ts:25]: similar note here - would be nice to either import with an `as getAdminServer`, or rename this to `getAdminServer`.

File: src/test/admin-server/index.ts [hunk 5]
```diff
@@ -1,25 +1,51 @@
...
 import { client } from '../../database/client';
+import { ContextManager } from '../../admin/context';
+import s3 from '../../aws/s3';
 
 // Export this separately so that it can be used in Apollo integration tests
 export const db = client();
 
+// Only using one value from MozillaAccessGroup enum in Pocket Shared Data
+export const COLLECTION_CURATOR_FULL =
+  'mozilliansorg_pocket_collection_curator_full';
```
Comment (jpetto) [inline src/test/admin-server/index.ts:19]: i don't know if this is a nit-pick or not - i feel like it's a little messy to keep this enum definition here. we're going to need the read only value as well for queries, and while i guess we could keep them both in here, it just feels...messy. would like to know how others feel.

cc @Herraj

File: src/test/admin-server/index.ts (hunk 1 above)
Comment (jpetto) [reply src/test/admin-server/index.ts:44]: i don't _think_ we need to worry about sentry in our test server. i'm okay leaving it out.

File: src/admin/resolvers/mutations.ts (hunk 2 above)
Comment (jpetto) [reply src/admin/resolvers/mutations.ts:73]: i like this suggestion. i think it makes the code more readable.

**Review by jpetto — CHANGES_REQUESTED** [review]
a few clean-up asks, but this is nice refactoring/updating work 😄 

i think @Herraj's suggestion on testing `undefined` headers is valid, and i like the idea of separate authorization only tests. i'd like to discuss that strategy in our planning meeting.

**Comment by jpetto** [comment]
@nina-py i added filling out the integration tests in this repo to our chose epic in jira 😄

File: src/admin/context.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
+import { PrismaClient } from '@prisma/client';
+import { S3 } from 'aws-sdk';
+import { Request } from 'express';
+import { client } from '../database/client';
+import s3 from '../aws/s3';
```
[the remark this answers is not in the document]
Comment (nina-py) [reply src/admin/context.ts:5]: Renamed to `s3service`! :) Not the best of names but at least more descriptive than just `s3`, right?

File: src/admin/resolvers/mutations.ts (hunk 2 above)
Comment (nina-py) [reply src/admin/resolvers/mutations.ts:73]: Done!

File: src/main.ts (hunk 3 above)
Comment (nina-py) [reply src/main.ts:10]: Fixed!

File: src/test/admin-server/index.ts (hunk 5 above)
Comment (nina-py) [reply src/test/admin-server/index.ts:19]: I agree, it was not a good spot for it and it was defined once more in the context. We don't really have any shared data here and the only `types.ts` file is for database resolvers, so I created a `src/shared/constants.ts` file and placed the constant there.

**Comment by nina-py** [comment]
Paired with @Herraj to add tests for when no authorisation headers are sent through - for example, when someone works with/tests the subgraph directly.

**Review by Herraj — APPROVED** [review]
🍐

**Review by jpetto — APPROVED** [review]
looks great!

**Comment by jpetto** [comment]
let's not merge this until we know all curators have this group. we don't want to push to prod and break collections.

**Pull request merged**